src/hotspot/share/opto/loopTransform.cpp

```diff
@@ -3072,8 +3086,14 @@ void PhaseIdealLoop::do_range_check(IdealLoopTree *loop, Node_List &old_new) {
     // Computed pre-loop limit can be outside of loop iterations range.
     pre_limit = (stride_con > 0) ? (Node*)new MinINode(pre_limit, orig_limit)
                                  : (Node*)new MaxINode(pre_limit, orig_limit);
-    register_new_node(pre_limit, pre_ctrl);
+    register_new_node(pre_limit, new_limit_ctrl);
   }
+  // new pre_limit can push Bool/Cmp/Opaque nodes down (when one of the eliminated condition has parameters that are not
+  // loop invariant in the pre loop.
+  set_ctrl(pre_opaq, new_limit_ctrl);
+  set_ctrl(pre_end->in(1), new_limit_ctrl);
+  set_ctrl(pre_end->cmp_node(), new_limit_ctrl);
```

Comment:
Just for better readability, I suggest to flip those such that it matches the top down order opaque -> cmp -> bool.
```suggestion
  set_ctrl(pre_end->cmp_node(), new_limit_ctrl);
  set_ctrl(pre_end->in(1), new_limit_ctrl);
```